Remotion/Web/Core/Resources/ResourcePathBuilderBasedStaticResourceCacheKeyProvider.cs

```diff
@@ -0,0 +1,266 @@
+  /// </summary>
+  /// <threadsafety static="true" instance="true" />
+  public class ResourcePathBuilderBasedStaticResourceCacheKeyProvider : IStaticResourceCacheKeyProvider, IDisposable
+  {
+    public interface IMappedResourcesPathLocator
+    {
+      string GetMappedResourcesPath ();
+    }
+
+    public class MappedResourcesPathLocator : IMappedResourcesPathLocator
+    {
+      private readonly IHttpContextProvider _httpContextProvider;
+      private readonly ResourceRoot _resourceRoot;
+
+      public MappedResourcesPathLocator (IHttpContextProvider httpContextProvider, ResourceRoot resourceRoot)
+      {
+        ArgumentUtility.CheckNotNull(nameof(httpContextProvider), httpContextProvider);
+        ArgumentUtility.CheckNotNull(nameof(resourceRoot), resourceRoot);
+
+        _httpContextProvider = httpContextProvider;
+        _resourceRoot = resourceRoot;
+      }
+
+      /// <inheritdoc />
+      public string GetMappedResourcesPath ()
+      {
+        var context = _httpContextProvider.GetCurrentHttpContext();
+        return context.Server.MapPath("~/" + _resourceRoot.Value);
+      }
+    }
+
+    public interface IResourceFileDetailsAppender
+    {
+      void AppendResourceFolderDetails (List<(string Name, long Length, long CreationTime, long LastWriteTime)> folderDetails, string resourceFolder, IReadOnlyCollection<ResourceType> resourceTypes);
+    }
+
+    public class ResourceFileDetailsAppender : IResourceFileDetailsAppender
+    {
+      /// <inheritdoc />
+      public void AppendResourceFolderDetails (List<(string Name, long Length, long CreationTime, long LastWriteTime)> folderDetails, string resourceFolder, IReadOnlyCollection<ResourceType> resourceTypes)
+      {
+        // Format of a resource folder is: '~/<assembly name>/<resource type>/...
+        foreach (var assemblyFolder in Directory.EnumerateDirectories(resourceFolder))
+        {
+          foreach (var resourceType in resourceTypes)
+          {
+            var resourceTypeDirectory = new DirectoryInfo(Path.Combine(assemblyFolder, resourceType.Name));
+            if (!resourceTypeDirectory.Exists)
+              continue;
+
+            foreach (var resourceFile in resourceTypeDirectory.EnumerateFiles("*", SearchOption.AllDirectories))
+            {
+              folderDetails.Add((resourceFile.FullName, resourceFile.Length, resourceFile.CreationTimeUtc.ToFileTimeUtc(), resourceFile.LastWriteTimeUtc.ToFileTimeUtc()));
+            }
+          }
+        }
+      }
+    }
+
+    private enum FileChangeType
+    {
+      Changed,
+      Created,
+      Deleted,
+      Renamed
+    }
+
+    private readonly IMappedResourcesPathLocator _mappedResourcesPathLocator;
+    private readonly IResourceFileDetailsAppender _resourceFileDetailsAppender;
+    private readonly IReadOnlyList<ResourceType> _resourceTypes;
+
+    private readonly DoubleCheckedLockingContainer<string?> _cacheKey;
+    private readonly HashSet<string> _resourceTypeLookup;
```

Comment:
Please make this also .net /.netfx contiional with .NET using the IReadOnlySet intead of HashSet. That way, we get threadsafety via compilercheck :)

Reply:
I don't follow what does `IReadOnlySet<T>` add here?

Reply:
It ensures that the values can only be changed in the ctor, i.e. the type ensures that we know the member is used in a non-modifyable way just by looking at the field declaration. This makes thread safety more easily verifyable.